src/coreclr/jit/morph.cpp

```diff
@@ -12631,6 +12646,90 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
     return cmp;
 }
 
+//------------------------------------------------------------------------
+// fgOptimizeRelationalComparisonWithFullRangeConst: optimizes a comparison operation.
+//
+// Recognizes "Always false" comparisons against various full range constant operands and morphs
+// them into zero.
+//
+// Arguments:
+//   cmp - the GT_LT/GT_GT tree to morph.
+//
+// Return Value:
+//   1. The unmodified "cmp" tree.
+//   2. A CNS_INT node containing zero.
+//   3. A GT_COMMA node containing side effects along with a CNS_INT node containing zero
+// Assumptions:
+//   The second operand is an integral constant or the first operand is an integral constant.
+//
+GenTree* Compiler::fgOptimizeRelationalComparisonWithFullRangeConst(GenTreeOp* cmp)
+{
+
+    int64_t lhsMin;
+    int64_t lhsMax;
+    if (cmp->gtGetOp1()->IsIntegralConst())
+    {
+        lhsMin = cmp->gtGetOp1()->AsIntConCommon()->IntegralValue();
+        lhsMax = lhsMin;
+    }
+    else
+    {
+        IntegralRange lhsRange = IntegralRange::ForNode(cmp->gtGetOp1(), this);
+        lhsMin                 = IntegralRange::SymbolicToRealValue(lhsRange.GetLowerBound());
+        lhsMax                 = IntegralRange::SymbolicToRealValue(lhsRange.GetUpperBound());
+    }
+
+    int64_t rhsMin;
+    int64_t rhsMax;
+    if (cmp->gtGetOp2()->IsIntegralConst())
+    {
+        rhsMin = cmp->gtGetOp2()->AsIntConCommon()->IntegralValue();
+        rhsMax = rhsMin;
+    }
+    else
+    {
+        IntegralRange rhsRange = IntegralRange::ForNode(cmp->gtGetOp2(), this);
+        rhsMin                 = IntegralRange::SymbolicToRealValue(rhsRange.GetLowerBound());
+        rhsMax                 = IntegralRange::SymbolicToRealValue(rhsRange.GetUpperBound());
+    }
+
+    genTreeOps op = cmp->gtOper;
+    if ((op != GT_LT) && (op != GT_LE))
+    {
+        op = GenTree::SwapRelop(op);
+        std::swap(lhsMin, rhsMin);
+        std::swap(lhsMax, rhsMax);
+    }
+
+    int64_t lefOpValue   = lhsMin;
+    int64_t rightOpValue = rhsMax;
+
+    if (cmp->IsUnsigned() && lefOpValue == -1 && lhsMax == -1)
+    {
+        rightOpValue = -1;
+    }
+
+    if ((op == GT_LT && lefOpValue == rightOpValue) || (op == GT_LE && lefOpValue > rightOpValue))
+    {
+        GenTree* ret = gtNewZeroConNode(TYP_INT);
+
+        if (gtTreeHasSideEffects(cmp, GTF_SIDE_EFFECT))
+        {
+            return cmp;
+        }
+
+        ret->SetVNsFromNode(cmp);
+
+        DEBUG_DESTROY_NODE(cmp);
+
+        INDEBUG(ret->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+
+        return ret;
+    }
```

Comment:
It doesn't seem right that this can only fold things into false. What about when the comparison is always true?

Reply:
I can't think of a case when the comparison is always true. There are 7 typical trees which I'm testing on (value types may vary but the trees are always generalized to the 7 cases):
1. When const is on the right and MinValue
example `v >= int.MinValue`:
```
    *  RETURN    int
    \--*  EQ        int
       +--*  LT        int
       |  +--*  LCL_VAR   int    V00 arg0
       |  \--*  CNS_INT   int    -0x80000000
       \--*  CNS_INT   int    0
```
2. When const is on the left and MinValue
example `int.MinValue <= v`
```
   *  RETURN    int
   \--*  EQ        int
      +--*  GT        int
      |  +--*  CNS_INT   int    -0x80000000
      |  \--*  LCL_VAR   int    V00 arg0

      \--*  CNS_INT   int    0
```
3. When const is on the right and MaxValue
example `v <= int.MaxValue`
```
 *  RETURN    int
 \--*  EQ        int
    +--*  GT        int
    |  +--*  LCL_VAR   int    V00 arg0
    |  \--*  CNS_INT   int    0x7FFFFFFF
    \--*  CNS_INT   int    0
```
4. When const is on the left and MaxValue
example `int.MaxValue >= v`
```
 *  RETURN    int
 \--*  EQ        int
    +--*  LT        int
    |  +--*  CNS_INT   int    0x7FFFFFFF
    |  \--*  LCL_VAR   int    V00 arg0
    \--*  CNS_INT   int    0
```
5. When const is ulong/uint and value is MinValue
example `v >= ulong.MinValue`
example `ulong.MinValue <= v`
```
*  RETURN    int
\--*  CNS_INT   int    1
```
6. When const is on the right and value type is ulong/uint and value is MaxValue
example `v <= ulong.MaxValue;`
```
 \--*  EQ        int
    +--*  GT        int
    |  +--*  LCL_VAR   long   V00 arg0
    |  \--*  CNS_INT   long   -1
    \--*  CNS_INT   int    0
```
7. When const is on the left and value type is ulong/uint and value is MaxValue
example `ulong.MaxValue >= v`
```
*  RETURN    int
\--*  EQ        int
   +--*  LT        int
   |  +--*  CNS_INT   long   -1
   |  \--*  LCL_VAR   long   V00 arg0
   \--*  CNS_INT   int    0
```

Reply:
So for example:
`v >= int.MinValue ` generates a tree that is equal to `return (v < int.MinValue) == false`
`int.MinValue <= v` generates a tree that is equal to `return (int.MinValue > v) == false`
However in case of ulong/uint:
`v <= ulong.MaxValue` generates a tree that is equal to `return (v > -1) == false
`ulong.MaxValue >= v` generates a tree that is equal to `return (-1 < v) == false
These trees looks odd to me because (v > -1) is going to be always true.
Correct me please if I'm wrong.

Reply:
One example would be:
`bool Foo(sbyte i) => i > -129;`

Reply:
We generally do not see `x >= y` much here because there is no such IL instruction, so that's why the more "common" pattern `i >= sbyte.MinValue` is reversed by Roslyn. But we can still see such IR if we introduce it ourselves, so I think it makes sense to handle it (and also the above case).

Reply:
Do you know how to generate the always true condition but with GT_LE? I'm considering should we be bothered with such a case at all.

Reply:
Given that you already have the intervals I think the actual check itself is so simple that there is no reason not to add it, it should not be more than a couple of lines.
[x0, x1] <= [y0, y1] is always true if x1 <= y0.

Reply:
I tried to apply this condition, which considering only LT for simplicity and it gave me 1700+ improvements during the spmi asmdiff run, which is suspicious :)
```
if (((op == GT_LT) && (lhsMin >= rhsMax)) || (((op == GT_LE) && (lhsMin > rhsMax))))
    {
        ret = gtNewZeroConNode(TYP_INT);
    }
    else if ((op == GT_LT) && rhsMax > lhsMax)
    {
        ret = gtNewOneConNode(TYP_INT);
    }
```
Then I tried to check which Op is const to make the condition more strict:
```
    //When lhs is constant
    *  RETURN    int
    \--*  LT       int
         +--*  CNS_INT   int    -129
          \--*  LCL_VAR   byte   V00 arg0
    else if ((op == GT_LT || (op == GT_LE)) && rhsMin > lhsMin && lhsMin == lhsMax)
    {
        ret = gtNewOneConNode(TYP_INT);
    }
    // When rhs is constant
      For cases like this:
      *  RETURN    int
      \--*  LT        int
          +--*  LCL_VAR   byte   V00 arg0
           \--*  CNS_INT   int    129

    else if ((op == GT_LT) && rhsMax > lhsMax && rhsMin == rhsMax)
    {
        ret = gtNewOneConNode(TYP_INT);
    }
```
It works fine and there are not so many improvements. What do you think of this?

Reply:
`[x0, x1] < [y0, y1]` is true if `x1 < y0`. So I think
`else if ((op == GT_LT) && rhsMax > lhsMax)`
should be
`else if ((op == GT_LT) && rhsMax < lhsMin)`.